dev/tools/localization/localizations_utils.dart

```diff
@@ -373,61 +373,46 @@ class $classNamePrefix$camelCaseName extends $superClass {''';
 
 /// Return the input string as a Dart-parseable string.
 ///
-/// The result tries to avoid character escaping:
+/// When [shouldEscapeDollar] is set to true, the
+/// result avoids character escaping, with the
+/// exception of the dollar sign:
 ///
 /// ```
+/// foo$bar = 'foo\$bar'
 /// foo => 'foo'
 /// foo "bar" => 'foo "bar"'
 /// foo 'bar' => "foo 'bar'"
 /// foo 'bar' "baz" => '''foo 'bar' "baz"'''
 /// foo\bar => 'foo\\bar'
-/// foo$bar => 'foo\$bar'
 /// ```
-String generateMessageString(String value) {
-  if (!value.contains("'"))
-    return "'$value'";
-  if (!value.contains('"'))
-    return '"$value"';
-  if (!value.contains("'''"))
-    return "'''$value'''";
-  if (!value.contains('"""'))
-    return '"""$value"""';
-
-  return value.split("'''")
-    .map(generateString)
-    // If value contains more than 6 consecutive single quotes some empty
-    // strings may be generated. The following map removes them.
-    .map((String part) => part == "''" ? '' : part)
-    .join(" \"'''\" ");
-}
-
-/// Return the input string as a Dart-parseable string.
 ///
-/// The result tries to avoid character escaping:
+/// When [shouldEscapeDollar] is set to false, the
+/// result tries to avoid character escaping:
 ///
 /// ```
-/// foo => 'foo'
-/// foo "bar" => 'foo "bar"'
-/// foo 'bar' => "foo 'bar'"
-/// foo 'bar' "baz" => '''foo 'bar' "baz"'''
-/// foo\bar => r'foo\bar'
+/// foo$bar => 'foo\\\$bar'
 /// ```
 ///
+/// [shouldEscapeDollar] is true by default.
+///
 /// Strings with newlines are not supported.
-String generateString(String value) {
+String generateString(String value, {bool shouldEscapeDollar = true}) {
```

Comment:
I think we can do this more simply, just by replacing each special character with a backslashed version. The only wrinkle is that the backslashes themselves have to be escaped, but only the ones in the original value string - not the ones we're adding. So:
```dart
String generateString(String value, { bool escapeDollar = true }) {
  assert(escapeDollar != null);
  assert(!value.contains('\n'));

  const backslash = '__BACKSLASH__';
  assert(!value.contains(backslash));
  value = value.replaceAll('\\', backslash);

  if (escapeDollar)
    value = value.replaceAll('\$', '\\\$');

  value = value
    .replaceAll("'", "\\'")
    .replaceAll('"', '\\"')
    .replaceAll(backslash, '\\\\');
  
  return "'$value'";
}
```